# <center>Practice 1. Analyzing Titanic passengers

**Fill in the missing code (`"You code here"`) and choose answers in a [web-form](https://forms.gle/oMuDa6HYPbPUc1GQ9).**

In [0]:
import pandas as pd
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [2]:
##### DOWNLOAD DATASET. ONLY IN COLAB !!!
!wget -P ./../data/ https://raw.githubusercontent.com/CenticMurcia/curso-ciencia-datos/master/data/titanic_train.csv

--2020-01-28 17:30:28--  https://raw.githubusercontent.com/CenticMurcia/curso-ciencia-datos/master/data/titanic_train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 60302 (59K) [text/plain]
Saving to: ‘./../data/titanic_train.csv’

titanic_train.csv   100%[===================>]  58.89K  --.-KB/s    in 0.01s   

2020-01-28 17:30:28 (4.00 MB/s) - ‘./../data/titanic_train.csv’ saved [60302/60302]



In [0]:
data = pd.read_csv('../data/titanic_train.csv', index_col='PassengerId')

**First 5 rows**

In [4]:
data.head()

Survived  Pclass  ... Cabin Embarked
PassengerId                    ...               
1                   0       3  ...   NaN        S
2                   1       1  ...   C85        C
3                   1       3  ...   NaN        S
4                   1       1  ...  C123        S
5                   0       3  ...   NaN        S

[5 rows x 11 columns]

In [5]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

In [6]:
data.describe(include=['object', 'bool'])

Name   Sex Ticket Cabin Embarked
count                                      891   891    891   204      889
unique                                     891     2    681   147        3
top     Laroche, Mr. Joseph Philippe Lemercier  male   1601    G6        S
freq                                         1   577      7     4      644

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [7]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

Survived  Pclass  ...            Cabin Embarked
PassengerId                    ...                          
119                 0       1  ...          B58 B60        C
259                 1       1  ...              NaN        C
300                 1       1  ...          B58 B60        C
312                 1       1  ...  B57 B59 B63 B66        C
378                 0       1  ...              C82        C

[5 rows x 11 columns]

**We can sort these people by Fare in descending order.**

In [8]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

Survived  Pclass  ...            Cabin Embarked
PassengerId                    ...                          
259                 1       1  ...              NaN        C
680                 1       1  ...      B51 B53 B55        C
738                 1       1  ...             B101        C
312                 1       1  ...  B57 B59 B63 B66        C
743                 1       1  ...  B57 B59 B63 B66        C

[5 rows x 11 columns]

**Let's create a new feature.**

In [0]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [0]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

**Another way is to do it with `apply`.**

In [0]:
data['Age_category'] = data['Age'].apply(age_category)

# 1. How many men/women were there onboard?
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- **577 men and 314 women**

In [14]:
# You code here
data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

# 2. How many men from second class were there onboard?

**Tip: Print the distribution of the `Pclass` feature with the `Sex` feature.**
- 104
- **108**
- 112
- 125

In [26]:
# You code here
len(data[(data['Pclass'] == 2) & (data['Sex'] == 'male')])

108

In [25]:
# Otra solución con tabla cruzada
pd.crosstab(data.Pclass, data.Sex)

Sex     female  male
Pclass              
1           94   122
2           76   108
3          144   347

# 3. What are median and standard deviation of `Fare`?
- median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [29]:
# You code here
print(data.Fare.median(), data.Fare.std())

14.4542 49.693428597180905


# 4. Is the mean age of survived people higher than people who died?
- Yes, the mean age of survived people is higher.
- No, the mean age of survived people is lower.

In [35]:
# You code here
data.pivot_table(['Age'], ['Survived'], aggfunc='mean')

Age
Survived       
0         30.63
1         28.34

In [43]:
# Otra forma
survived = data[data.Survived == 1].Age.mean()
died     = data[data.Survived == 0].Age.mean()
print(survived, died)

28.343689655172415 30.62617924528302


# 5. What are the percentages of survived passengers for:
## a) younger than 30 y.o.
## b) older than 60 y.o.

- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [51]:
# You code here
young = data[ data.Age < 30];
old   = data [ data.Age > 60];

print("Surviving young people: ", 100 * young.Survived.mean())
print("Surviving old people: ", 100 * old.Survived.mean())

Surviving young people:  40.625
Surviving old people:  22.727272727272727


# 6. Is that true that women survived more frequently than men? What are the percentages?
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women

In [57]:
# You code here
women = data[data['Sex']=="female"]
men   = data[data['Sex']=="male"]

print("Survived women:", round(100 * women.Survived.mean(), 1), "%")
print("Survived men:  ", round(100 * men.Survived.mean(), 1), "%")

Survived women: 74.2 %
Survived men:   18.9 %


# 7. What's the most popular first name among male passengers?
- Charles
- Thomas
- William
- John

In [67]:
def getNameFunction(fullname):
    return fullname.split(",")[1].split(" ")[2]

# You code here
data[data.Sex == 'male'].Name.apply(getNameFunction).value_counts().head(1)

William    35
Name: Name, dtype: int64

# 8. How is average age for men/women dependent on `Pclass`?
### Choose all correct statements:
- On average, men of 1 class are older than 40
- On average, women of 1 class are older than 40
- Men of all classes are on average older than women of the same class
- On average, passengers ofthe first class are older than those of the 2nd class who are older than passengers of the 3rd class

In [70]:
# You code here
pd.crosstab(data.Pclass, data.Sex, values=data.Age, aggfunc='mean', margins=True)

Sex     female   male    All
Pclass                      
1        34.61  41.28  38.23
2        28.72  30.74  29.88
3        21.75  26.51  25.14
All      27.92  30.73  29.70